rsyslog/undefined_field/undefined_field.go

```diff
@@ -0,0 +1,392 @@
+ *    "CDM_EXTRA_KEEP_FIELDS": "",
+ *    "CDM_UNDEFINED_NAME": "undefined",
+ *    "CDM_KEEP_EMPTY_FIELDS": "",
+ *    "CDM_UNDEFINED_TO_STRING": "false",
+ *    "CDM_UNDEFINED_DOT_REPLACE_CHAR": "UNUSED",
+ *    "CDM_UNDEFINED_MAX_NUM_FIELDS": "-1",
+ *    "MERGE_JSON_LOG": "false",
+ *    "UNDEFINED_DEBUG": "false"
+ *  }
+ */
+
+package main
+
+import (
+	"encoding/json"
+	"bufio"
+	"strconv"
+	"strings"
+	"fmt"
+	"os"
+)
+
+const (
+	initial_undefined_debug            = false
+	initial_merge_json_log             = false
+	initial_use_undefined              = false
+	initial_undefined_name             = "undefined"
+	initial_undefined_to_string        = false
+	initial_default_keep_fields        = "CEE,time,@timestamp,aushape,ci_job,collectd,docker,fedora-ci,file,foreman,geoip,hostname,ipaddr4,ipaddr6,kubernetes,level,message,namespace_name,namespace_uuid,offset,openstack,ovirt,pid,pipeline_metadata,rsyslog,service,systemd,tags,testcase,tlog,viaq_msg_id"
+	initial_extra_keep_fields          = ""
+	initial_keep_empty_fields          = ""
+	initial_undefined_max_num_fields   = -1
+	initial_undefined_dot_replace_char = "UNUSED"
+	initial_logging_file_path          = "/var/log/rsyslog/rsyslog.log"
+	undefined_config                   = "/var/lib/rsyslog.pod/undefined.json"
+	noChanges                          = "{}"
+)
+
+var (
+	undefined_debug bool
+	merge_json_log bool
+	use_undefined bool
+	keep_fields map[string]string
+	keep_empty_fields map[string]string
+	undefined_name string
+	undefined_to_string bool
+	undefined_dot_replace_char string
+	undefined_max_num_fields int64
+	logfile *os.File
+	noaction		  = false
+	replacer		  = &strings.Replacer{}
+)
+
+func getMapStringValue(m map[string]interface{}, key string) (string, bool) {
+	if val, ok := m[key]; ok {
+		return val.(string), ok
+	} else {
+		return "", ok
+	}
+}
+
+func onInit() {
+	// opening the rsyslog log file
+	logging_file_path := initial_logging_file_path
+	if eval := os.Getenv("LOGGING_FILE_PATH"); eval != "" {
+		logging_file_path = eval
+	}
+	if f, err := os.OpenFile(logging_file_path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644); err == nil {
+		logfile = f
+	} else {
+		panic(fmt.Errorf("Could not open file [%s]: [%v]", logging_file_path, err))
+	}
+
+	// initializing
+	undefined_debug = initial_undefined_debug
+	merge_json_log = initial_merge_json_log
+	use_undefined = initial_use_undefined
+	default_keep_fields := initial_default_keep_fields
+	extra_keep_fields := initial_extra_keep_fields
+	undefined_name = initial_undefined_name
+	tmp_keep_empty_fields := initial_keep_empty_fields
+	undefined_to_string = initial_undefined_to_string
+	undefined_dot_replace_char = initial_undefined_dot_replace_char
+	undefined_max_num_fields = initial_undefined_max_num_fields
+
+	if config, err := os.Open(undefined_config); err == nil {
```

Comment:
I think this is ok for now, but fyi the simpler way would have been to define a config struct and use the built-in json to struct parsing e.g.
```go
type UndefinedConfig struct {
  bool Debug `json:"UNDEFINED_DEBUG"`
  bool Merge_json_log `json:"MERGE_JSON_LOG"`
  bool Use_undefined `json:"CDM_USE_UNDEFINED"`
  string Default_keep_fields `json:"CDM_DEFAULT_KEEP_FIELDS"`
  ...
  int Undefined_max_num_fields `json:"CDM_UNDEFINED_MAX_NUM_FIELDS"`
}
...
var UndefinedConfig undefined_config_obj
// not sure how to convert the file object config to an array of bytes
json.Unmarshal([]byte(config.read()), &undefined_config_obj)
// or maybe dec = json.NewDecoder(config)
// dec.Decode(&undefined_config_obj)
fmt.Println("undefined_debug %v", undefined_config_obj.Debug)
```
see https://blog.golang.org/json-and-go and https://www.sohamkamani.com/blog/2017/10/18/parsing-json-in-golang/

Reply:
Ok.  Let me apply your patch and rerun the test...  Thanks!